cookbooks/cdo-varnish/libraries/helpers.rb

```diff
@@ -0,0 +1,117 @@
+# Various Ruby helper methods to generate Varnish VCL from the base configuration.
+
+# Basic regex matcher for the optional query part of a URL.
+QUERY_REGEX = "(\\?.*)?$"
+
+def path_to_regex(path)
+  path = "^/#{path.sub(/^\//,'')}"
+  if path[-1] == '*'
+    path.slice!(-1)
+  else
+    path = path + QUERY_REGEX
+  end
+  path
+end
+
+# Varnish uses 'bereq' in the 'response' section, 'req' otherwise.
+def req(section)
+  section == 'response' ? 'bereq' : 'req'
+end
+
+# Returns a regex-matcher string based on an array of filename extensions.
+def extensions_to_regex(exts)
+  return [] if exts.empty?
+  ["(?i)(#{exts.join('|').gsub('.','\.')})#{QUERY_REGEX}"]
+end
+
+# Returns a regex-conditional string fragment based on the provided behavior.
+# In the 'proxy' section, ignore extension-based behaviors (e.g., *.png).
+def paths_to_regex(path_config, section)
+  path_config = [path_config] unless path_config.is_a?(Array)
+  extensions, paths = path_config.partition{ |path| path[0] == '*' }
+  elements = paths.map{|path| path_to_regex(path)}
+  elements = extensions_to_regex(extensions.map{|x|x.sub(/^\*/,'')}) + elements unless section == 'proxy'
+  elements.empty? ? 'false' : elements.map{|el| "#{req(section)}.url ~ \"#{el}\""}.join(' || ')
+end
+
+# Generates the logic string for the specified behavior.
+def process_behavior(behavior, app, section)
+  if section == 'proxy'
+    process_proxy(behavior['proxy'] || app)
+  else
+    process_cookies(behavior['cookies'], section)
+  end
+end
+
+# Returns the cookie-filter string for a given 'cookies' behavior.
+def process_cookies(cookies, section)
+  if section == 'request'
+    cookies = ['NO_CACHE'] if cookies == 'none'
+    "cookie.filter_except(\"#{cookies.join(',')}\");"
+  else
+    cookies == 'none' ?
+      'unset beresp.http.set-cookie;' :
+      ''
+  end
+end
+
+# Returns the backend-redirect string for a given proxy.
+def process_proxy(proxy)
+  "set req.backend_hint = #{proxy}.backend();"
+end
+
+# Returns the canonical hostname based on the current Chef node's name/environment.
+# Keep in sync with CDO.canonical_hostname in deployment.rb.
+def canonical_hostname(domain)
```

Comment:
This should also be unit tested to make sure we cover all of the possible domains of interest.